File: Graphs/graph-prim.java
// Prim's Algorithm
// Time: O(E log V)
// Space: O(V + E)

import java.util.*;

public class Prim {
    static class Edge {
        int dest, weight;
        Edge(int d, int w) { dest = d; weight = w; }
    }

    public static List<int[]> prim(List<List<Edge>> graph, int start) {
        int V = graph.size();
        boolean[] visited = new boolean[V];
        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a[1]));
        List<int[]> MST = new ArrayList<>();

        pq.add(new int[]{start, 0, -1}); // vertex, weight, parent

        while (!pq.isEmpty()) {
            int[] curr = pq.poll();
            int u = curr[0], w = curr[1], parent = curr[2];
            if (visited[u]) continue;
            visited[u] = true;

            if (parent != -1) MST.add(new int[]{parent, u, w});

            for (Edge e : graph.get(u)) {
                if (!visited[e.dest]) {
                    pq.add(new int[]{e.dest, e.weight, u});
                }
            }
        }
        return MST;
    }
}

File: Graphs/graph-prim.py
# Prim's Algorithm
# Time: O(E log V)
# Space: O(V + E)
import heapq

def prim(graph, start=0):
    MST = []
    visited = set()
    min_heap = [(0, start, None)]  # (weight, vertex, parent)
    
    while min_heap:
        weight, u, parent = heapq.heappop(min_heap)
        if u in visited:
            continue
        visited.add(u)
        if parent is not None:
            MST.append((parent, u, weight))
        
        for v, w in graph[u]:
            if v not in visited:
                heapq.heappush(min_heap, (w, v, u))
    return MST
